src/controller/CHIPDeviceController.cpp

```diff
@@ -293,17 +291,18 @@ CHIP_ERROR DeviceController::Shutdown()
     ChipLogDetail(Controller, "Shutting down the controller");
 
 #if CONFIG_DEVICE_LAYER
-    // Start by shutting down the PlatformManager.  This will ensure, with
-    // reasonable synchronization, that we stop processing of incoming messages
-    // before doing any other shutdown work.  Otherwise we can end up trying to
-    // process incoming messages in a partially shut down state, which is not
-    // great at all.
-    ReturnErrorOnFailure(DeviceLayer::PlatformMgr().Shutdown());
+    //
+    // We don't call the PlatformMgr's() Shutdown() API since that is a thread-safe API,
```

Comment:
Is this still needed after #7430 ?

Reply:
Somebody is misunderstanding something here.  It could well be me, so apologies if that's the case.

But on POSIX at least, PlatformMgr().Shutdown() is pthread_join() to the event loop thread.  Not only do you not need the lock for this, you must NOT have the lock held when calling this.

Instead, by clearing the should-run flag and performing the pthread_join, you're safely pending on exit of the event loop thread.  Once this occurs, there is no longer have another thread to contend with and subsequent shutdown can occur without the lock.

Reply:
OK, it was me slightly misunderstanding.  Yes, I see now that POSIX `_StopEventLoopTask` has locking, and only holds the lock for this:

```
    mShouldRunEventLoop.store(false, std::memory_order_relaxed);
    if (mChipTask)    {
        SystemLayer.WakeSelect();
```

It would be nice to leverage what I think was the original intent, using atomics, and not have to hold the lock here.  But that's an aside.

I have a larger question though.  On POSIX at least:(1) the code can't safely call these subsequent shutdown and delete operations until the event loop thread is pthread_joined, and (2) it is obviously impossible to pthread_join the event loop thread from the event loop thread itself.

This all means that, at least on POSIX, both Controller shutdown and Platform shutdown (pthread_join) must be called from another thread.  So at least on POSIX, it's unclear how the separation of `TeardownChipStack` accomplishes anything.  How does the caller leverage this refactor to make shutdown more sensible?  It still isn't possible to dispatch these shutdown operations to the event loop itself.

Reply:
Right, so the current situation seems to be that Controller and Platform shutdown needs to happen from some not-the-message-handling thread thread.  My general point here is that if the first thing Controller shutdown does is Platform shutdown after that we are the only thread left and we should not need any locking here...

And yes, I agree we need to not be holding the lock at all while calling Platform shutdown.

Reply:
Also, it is _very_ possible that I am missing something here.  This stuff is not very straightforward.  :(

Reply:
> The problem is that the call to WakeSelect() touches the critical section within the SystemLayer, which is accessed by the CHIP thread as well when it comes out of select in HandleSelectResult(). This was flagged by tsan when I ran it without those locks (which is why I added them in the first place :P).
> 
> >

OK, I see then.  Yes, this is a situation where whatever function ends up with the call chain to pthread_join also needs to internally lock because of the above.  So you need the wrapper with locking and the separate, non-locking stack teardown function that this also calls.

I'm sure we could waste tons of time coming up with alternative names for these functions, but suffice to say I now understand why this is structured this way and see why it needs to be.

Reply:
I haven't though much about naming, so feel free to disregard this.  But since input was solicited:

> I'm treating DeviceController::Shutdown like any other method in that class, i.e it expects to be called with external synchronization.
> 
> Thoughts?

Yeah, this seems like a good thing.  Consistency is good.

Perhaps along these lines, the following could also help:

* revert naming for what is now `TeardownChipStack` back to `Shutdown`
* retain `StopEventLoopTask` as strictly separate from `Shutdown`

Here's my rationale.  `PlatformMgr().Shutdown` is now a special case that cannot be executed from within the event loop itself on platforms that spawn a separate event loop thread.  So our desire for abstraction and uniformity with a Shutdown like this just doesn't work; it has different calling rules on different platforms.

Instead, we can keep `PlatformMgr().Shutdown()` as the collection of all things that don't have this problem (currently the things encapsulated by `TeardownChipStack`).

Then, on platforms where a separate event loop thread is spawned, you will have to call `StopEventLoopTask` before `PlatformMgr().Shutdown()`.  I think this is OK because it is already unworkable for platforms like this to be unaware of the additional constraints arising from their separate event loop thread.

Reply:
I had debated that very same idea, and the only reason I had initially dismissed it was my hesitation to alter dramatically some of those contracts in Shutdown since it would increase my scope of work :)

I think I was over-reaching in trying to provide too much flexibility with both the variant that 'just tears down the stack' vs. one that does both 'teardown + thread termination'. 

Pivoting `Shutdown()` to do that I think would also align nicely with DeviceController::Shutdown()'s expectations (and pretty much, every other Shutdown API in all the other objects), while providing a separate API to actually shutdown the thread. This makes it also quite conscious the intention a developer has to make, vs. making it implicit and hidden this fairly consequential side-effect of thread termination.

Will make the change! (and consequently, update other callers to Shutdown...)

Reply:
Hmm, I've run into a minor snag with this approach.

Take minimal-mdns example application for example (examples/minimal-mdns/client.cpp) - it adopts a model where it runs the application logic and the stack on the same thread (the default main thread the process starts up in).

It schedules a timer to eventually terminate the application - this callback from the timer runs on the same thread as the application.

The application eventually blocks on `RunEventLoop()`. 

In this case, the termination is invoked by  a call to `PlatformMgr().Shutdown()`. Previously, this used to set the `mShouldRunEventLoop` to false to correctly cause the event loop to exit and eventually, result in the application shutting down.

I don't have that in my updated logic/scope for PlatformMgr().Shutdown().

What should the app do? Call `StopEventLoopTask()`? That would in this correctly, correctly not actually try to pthread_join the chip task since mChipTask within the POSIX implementation will not be set. But this would be calling that from within the CHIP thread. It won't deadlock (I've since moved the LockChipStack() and Unlock() calls within _StopEventLoopTask() to inside the if {} statement).

However, what if there was an implementation that decided to create a separate CHIP task, and then from within that, call StopEventLoopTask? It would deadlock.

So not sure how to proceed here @msandstedt. I would like to come out of this with a sane API that developers don't have to think too much about.

Suggestions are welcome.

Reply:
One option is to rename `StopEventLoopTask` to `StopEventLoop`, with the by-product that it will also terminate the CHIP thread if called from a different thread context.

Reply:
Minor nit: this needs to use `pthread_equal` for the comparison to pthread_selfI().  Other than that, seems like a good solution and finally gets us to the uniformity we want to have.

I was trying to figure out if there are any safety issues around the mChipTask variable.  I don't think there are any.  If there is an attempt to call `_StopEventLoopTask` simultaneously from both the non-event-loop thread and the event-loop thread (which one should not do), I believe this code would still work properly and often return `ESRCH` from pthread_join (that is, the improper use would not lead to undefined behavior and would often produce a sane error).  

As to reads of `mChipTask` not being atomic, it appears that in this particular case, no problems can arise from that.  I believe this to be the case because the non-event-loop thread is the only writer and the event-loop thread is strictly a reader, and the writer's writes are implicitly synchronized on `pthread_create` and `pthread_join` calls.

Looks great to me.

Reply:
Awesome! Will make the changes soon.

Reply:
Another small wrinkle!
A new example app causing a wrinkle: TestMdns.cpp

It spawns a thread to run the CHIP event loop on its own (doesn't call into StartEventLoopTask), and the main thread, arms a timer and when it fires, tears down the stack.

When the timer fires, calling StopEventLoopTask() just sets the mShouldRunEventLoop flag to false, but doesn't actually wait for the event queue to stop.

So this triggers a race condition detection error by tsan (good job tsan!), since the calling thread continues ahead to calling Shutdown, which races accessing the SDK with the CHIP thread that is still doing work and hasn't actually gotten around to stopping the event queue.

StopEventLoopTask() doesn't wait for the event loop to stop (by waiting on pthread_join) since in this case, a thread was created externally outside of the PlatformMgr's awareness.

Suggestion: Set mChipTask to point to the thread that is running the event loop when _RunEventLoop() is called, and set a special flag that indicates whether it's 'externally managed' or 'internally managed' thread.

In this case, it would be externally managed.

In StopEventLoopTask, we check to see if pthread_sef() is != mChipTask, if it's not the same, then check if it's externally or internally managed. If internal, do what we do today (signal wake, pthread_join).
If external, signal, then wait on a cond var till the event queue stops (which in turn, would signal back).

We shouldn't try to assume the thread will terminate if externally managed, hence why I think we shouldn't do pthread_join

Reply:
I've made the requisite fixes to the POSIX implementation, and had to do some fairly major surgery there.

@msandstedt please check it out.